cmd/jwt_parser.go

```diff
@@ -0,0 +1,173 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+// This file is a re-implementation of the original code here with some
+// additional allocation tweaks reproduced using GODEBUG=allocfreetrace=1
+// original file https://github.com/dgrijalva/jwt-go/blob/master/parser.go
+// borrowed under MIT License https://github.com/dgrijalva/jwt-go/blob/master/LICENSE
+
+import (
+	"crypto"
+	"crypto/hmac"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"sync"
+
+	jwtgo "github.com/dgrijalva/jwt-go"
+	jsoniter "github.com/json-iterator/go"
+)
+
+// SigningMethodHMAC - Implements the HMAC-SHA family of signing methods signing methods
+// Expects key type of []byte for both signing and validation
+type SigningMethodHMAC struct {
+	Name string
+	Hash crypto.Hash
+}
+
+// Specific instances for HS256 and company
+var (
+	SigningMethodHS256 *SigningMethodHMAC
+	SigningMethodHS384 *SigningMethodHMAC
+	SigningMethodHS512 *SigningMethodHMAC
+)
+
+var hmacSigners = []*SigningMethodHMAC{
+	{"HS256", crypto.SHA256},
+	{"HS384", crypto.SHA384},
+	{"HS512", crypto.SHA512},
+}
+
+type keyFunc func(*MapClaims) ([]byte, error)
+
+// ParseWithClaims - parse the token string, valid methods.
+func ParseWithClaims(tokenStr string, claims *MapClaims, fn keyFunc) error {
+	bufp := spool.Get().(*[]byte)
+	defer spool.Put(bufp)
+
+	signer, err := ParseUnverified(tokenStr, claims, *bufp)
+	if err != nil {
+		return err
+	}
+
+	// Lookup key
+	if fn == nil {
+		// keyFunc was not provided.  short circuiting validation
+		return jwtgo.NewValidationError("no Keyfunc was provided.", jwtgo.ValidationErrorUnverifiable)
+	}
+
+	key, err := fn(claims)
+	if err != nil {
+		return err
+	}
+
+	// Validate Claims
+	if err = claims.Valid(); err != nil {
+		// If the Claims Valid returned an error, check
+		// if it is a validation error, If it was another
+		// error type, create a jwtgo.ValidationError with
+		// a generic ClaimsInvalid flag set
+		return err
+	}
+
+	i := strings.LastIndex(tokenStr, ".")
+
+	n, err := base64Decode(tokenStr[i+1:], *bufp)
+	if err != nil {
+		return err
+	}
+
+	hasher := hmac.New(signer.Hash.New, key)
+	hasher.Write([]byte(tokenStr[:i]))
+	if !hmac.Equal((*bufp)[:n], hasher.Sum(nil)) {
+		return errAuthentication
+	}
+
+	return nil
+}
+
+var spool = sync.Pool{
+	New: func() interface{} {
+		buf := make([]byte, 1024)
+		return &buf
+	},
+}
```

Comment:
Could you declare this var before its first usage in this file?

Reply:
why? this is the more idiomatic usage of sync.Pool @donatello

Reply:
I just meant `spool` can be declared before the Parse function that uses - just a small readability improvement.

Reply:
Pools are global either they are a var or declared in `init()` they are not to be treated as global variables inside a function.